File: packages/react-native/android/src/main/java/dev/sinua/reactnative/SinuaVoiceModule.kt
package dev.sinua.reactnative

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * The React Native bridge for native voice sources (src/voice.ts). It only forwards
 * to [VoiceRegistry] -- create / connect / disconnect / release and the
 * `getCredential` round trip -- and streams the registry's events to JS as one
 * `sinua-voice` event carrying the source's id.
 *
 * Credentials arrive in `create` (or per request) and go straight into the source;
 * this class keeps none of them and logs none of them.
 */
class SinuaVoiceModule(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    override fun getName() = "SinuaVoice"

    init {
        VoiceRegistry.onEvent = { id, event, payload ->
            val body = Arguments.createMap()
            body.putString("id", id)
            body.putString("event", event)
            for ((k, v) in payload) if (v is String) body.putString(k, v)
            reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                .emit("sinua-voice", body)
        }
    }

    /** RN's NativeEventEmitter calls these; the emitter here is the device event one. */
    @ReactMethod fun addListener(eventName: String) = Unit

    @ReactMethod fun removeListeners(count: Int) = Unit

    @ReactMethod
    fun create(config: ReadableMap, promise: Promise) {
        try {
            VoiceRegistry.create(reactContext, config)
            promise.resolve(config.getString("id"))
        } catch (e: Exception) {
            promise.reject("sinua_voice_create", e.message, e)
        }
    }

    @ReactMethod
    fun connect(id: String, promise: Promise) {
        scope.launch {
            try {
                // Sources are main-thread objects (the native contract), and connect() returns
                // once the socket is open; later failures arrive as an `error` event instead.
                withContext(Dispatchers.Main) { VoiceRegistry.connect(id) }
                promise.resolve(null)
            } catch (e: Exception) {
                promise.reject("sinua_voice_connect", e.message, e)
            }
        }
    }

    @ReactMethod
    fun disconnect(id: String) {
        scope.launch { VoiceRegistry.disconnect(id) }
    }

    @ReactMethod
    fun release(id: String) {
        scope.launch { VoiceRegistry.release(id) }
    }

    @ReactMethod
    fun provideCredential(requestId: String, credential: String?, error: String?) {
        VoiceRegistry.provideCredential(requestId, credential, error)
    }
}

File: packages/react-native/ios/SinuaVoiceModule.swift
import Foundation
import React

/// The React Native bridge for native voice sources (src/voice.ts). It only
/// forwards to `VoiceRegistry`: create / connect / disconnect / release, plus the
/// `getCredential` round trip, and streams the registry's events to JS as one
/// `sinua-voice` event carrying the source's id.
///
/// Credentials arrive in `create` (or per request) and go straight into the
/// source; this class keeps none of them and logs none of them.
@objc(SinuaVoice)
class SinuaVoice: RCTEventEmitter {
    private var listening = false

    override static func requiresMainQueueSetup() -> Bool { true }

    override func supportedEvents() -> [String] { ["sinua-voice"] }

    override func startObserving() {
        listening = true
        Task { @MainActor in
            VoiceRegistry.shared.onEvent = { [weak self] id, event, payload in
                guard let self, self.listening else { return }
                var body: [String: Any] = payload
                body["id"] = id
                body["event"] = event
                self.sendEvent(withName: "sinua-voice", body: body)
            }
        }
    }

    override func stopObserving() {
        listening = false
    }

    @objc(create:resolver:rejecter:)
    func create(_ config: NSDictionary, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        let dict = config as? [String: Any] ?? [:]
        Task { @MainActor in
            do {
                try VoiceRegistry.shared.create(config: dict)
                resolve(dict["id"] as? String ?? "")
            } catch {
                reject("sinua_voice_create", error.localizedDescription, nil)
            }
        }
    }

    @objc(connect:resolver:rejecter:)
    func connect(_ id: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        Task { @MainActor in
            do {
                try await VoiceRegistry.shared.connect(id: id)
                resolve(NSNull())
            } catch {
                // Failures after a successful connect go to JS as an `error` event instead.
                reject("sinua_voice_connect", error.localizedDescription, nil)
            }
        }
    }

    @objc(disconnect:)
    func disconnect(_ id: String) {
        Task { @MainActor in VoiceRegistry.shared.disconnect(id: id) }
    }

    @objc(release:)
    func release(_ id: String) {
        Task { @MainActor in VoiceRegistry.shared.release(id: id) }
    }

    @objc(provideCredential:credential:error:)
    func provideCredential(_ requestId: String, credential: NSString?, error: NSString?) {
        Task { @MainActor in
            VoiceRegistry.shared.provideCredential(requestId: requestId, credential: credential as String?, error: error as String?)
        }
    }
}
